### PR TITLE
GLT-2252: auto-sort by sample box column

### PR DESCRIPTION
**Pull request by djcooke**
when propagating libraries from the box page

### CODE REVIEW

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java [hunk 1]
```diff
@@ -703,8 +726,9 @@ private void writeLibraryConfiguration(ObjectMapper mapper, ObjectNode config) {
 
   @RequestMapping(value = "/bulk/propagate", method = RequestMethod.GET)
   public ModelAndView propagateFromSamples(@RequestParam("ids") String sampleIds, @RequestParam("replicates") int replicates,
-      ModelMap model) throws IOException, MisoNamingException {
-    return libraryBulkPropagateBackend.propagate(sampleIds, replicates, model);
+      @RequestParam(name = "sort", required = false) String sort, ModelMap model) throws IOException, MisoNamingException {
```
Comment (sqwci) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java]: MINOR Remove the declaration of thrown exception 'uk.ac.bbsrc.tgac.miso.core.exception.MisoNamingException', as it cannot be thrown from method's body. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java [hunk 2]
```diff
@@ -686,12 +700,21 @@ protected LibraryDto createDtoFromParent(Sample item) {
...
-  };
+
+    public ModelAndView propagate(String idString, int replicates, String sort, ModelMap model) throws IOException {
+      this.sort = sort;
+      return propagate(idString, replicates, model);
+    }
+
+  }
 
   private void writeLibraryConfiguration(ObjectMapper mapper, ObjectNode config) {
```
Comment (sqwci) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java]: MAJOR Remove this unused method parameter "mapper". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1172)

File: miso-web/src/main/webapp/scripts/hot_sample.js [hunk 3]
```diff
@@ -584,93 +585,95 @@ HotTarget.sample = (function() {
...
-                    }
-                  };
+      return [{
+        name: "Edit",
+        action: function(samples) {
+
+          if (samples.some(function(sample) {
+            return sample.sampleClassId;
+          }) && !Constants.isDetailedSample) {
+            alert("There's detailed samples, but MISO is not configured for this.");
```
Comment (sqwci) [inline miso-web/src/main/webapp/scripts/hot_sample.js]: MINOR Remove this usage of alert(...). [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS1442)

File: miso-web/src/main/webapp/scripts/hot_sample.js [hunk 4]
```diff
@@ -584,93 +585,95 @@ HotTarget.sample = (function() {
...
+            return;
+          }
+
+          var classes = getSampleClasses(samples);
+          var classesAliases = Utils.array.deduplicateString(classes.map(function(sampleClass) {
+            return sampleClass.alias;
+          }));
+          if (classesAliases.length > 1) {
+            alert("You have selected samples of classes " + classesAliases.join(" & ") + ". Please select samples from only one class.");
```
Comment (sqwci) [inline miso-web/src/main/webapp/scripts/hot_sample.js]: MINOR Remove this usage of alert(...). [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS1442)

File: miso-web/src/main/webapp/scripts/hot_sample.js [hunk 5]
```diff
@@ -584,93 +585,95 @@ HotTarget.sample = (function() {
...
-                  return !!x;
-                }), function(result) {
-                  (result.target || targets[0]).action(result.replicates);
-                });
               });
             }
-          }, HotUtils.printAction('sample'), HotUtils.spreadsheetAction('/miso/rest/sample/spreadsheet', Constants.sampleSpreadsheets),
+
+            if (targets.length == 0) {
+              alert("No propagation is possible from the samples.");
```
Comment (sqwci) [inline miso-web/src/main/webapp/scripts/hot_sample.js]: MINOR Remove this usage of alert(...). [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS1442)

**Comment by sqwci** [comment]
SonarQube analysis reported 30 issues
* ![BLOCKER][BLOCKER] 3 blocker
* ![CRITICAL][CRITICAL] 9 critical
* ![MAJOR][MAJOR] 7 major
* ![MINOR][MINOR] 11 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![BLOCKER][BLOCKER] [hot.js#L336](https://github.com/TGAC/miso-lims/blob/4f62d8cb6f3638abc579ce43261ec9d63afc88ce/miso-web/src/main/webapp/scripts/hot.js#L336): Add the "let", "const" or "var" keyword to this declaration of "prop" to make it explicit. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2703)
1. ![BLOCKER][BLOCKER] [hot.js#L381](https://github.com/TGAC/miso-lims/blob/4f62d8cb6f3638abc579ce43261ec9d63afc88ce/miso-web/src/main/webapp/scripts/hot.js#L381): Add the "let", "const" or "var" keyword to this declaration of "needsRender" to make it explicit. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2703)
1. ![BLOCKER][BLOCKER] [hot.js#L1005](https://github.com/TGAC/miso-lims/blob/4f62d8cb6f3638abc579ce43261ec9d63afc88ce/miso-web/src/main/webapp/scripts/hot.js#L1005): Add the "let", "const" or "var" keyword to this declaration of "HotTarget" to make it explicit. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2703)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L550](https://github.com/TGAC/miso-lims/blob/4f62d8cb6f3638abc579ce43261ec9d63afc88ce/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L550): Define a constant instead of duplicating this literal "title" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L577](https://github.com/TGAC/miso-lims/blob/4f62d8cb6f3638abc579ce43261ec9d63afc88ce/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L577): Define a constant instead of duplicating this literal "library" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L611](https://github.com/TGAC/miso-lims/blob/4f62d8cb6f3638abc579ce43261ec9d63afc88ce/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L611): Define a constant instead of duplicating this literal "Libraries" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L819](https://github.com/TGAC/miso-lims/blob/4f62d8cb6f3638abc579ce43261ec9d63afc88ce/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L819): Define a constant instead of duplicating this literal "Dilutions" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L834](https://github.com/TGAC/miso-lims/blob/4f62d8cb6f3638abc579ce43261ec9d63afc88ce/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L834): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L857](https://github.com/TGAC/miso-lims/blob/4f62d8cb6f3638abc579ce43261ec9d63afc88ce/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L857): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L919](https://github.com/TGAC/miso-lims/blob/4f62d8cb6f3638abc579ce43261ec9d63afc88ce/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L919): Define a constant instead of duplicating this literal "_POOL" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'
[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: miso-web/src/main/webapp/scripts/hot_library.js [hunk 6]
```diff
@@ -155,7 +162,17 @@ HotTarget.library = (function() {
...
-
+    onLoad: function(config, table) {
+      if (config.sort) {
+        var sortOption = sortOptions[config.sort];
+        var sortColIndex = table.propToCol(sortOption.sortColumn);
+        var rowCount = table.countRows();
+        for (var i = 0; i < rowCount; i++) {
+          table.setCellMeta(i, sortColIndex, 'sortFunction', sortOption.sortFunction);
+        }
+        table.sort(sortColIndex);
```
Comment (callunity) [inline miso-web/src/main/webapp/scripts/hot_library.js]: Could extract the code in `hot.js#396-402` into a function and use that here as well...

**Comment by sqwci** [comment]
SonarQube analysis reported 28 issues
* ![BLOCKER][BLOCKER] 3 blocker
* ![CRITICAL][CRITICAL] 9 critical
* ![MAJOR][MAJOR] 6 major
* ![MINOR][MINOR] 10 minor

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![BLOCKER][BLOCKER] [hot.js#L341](https://github.com/TGAC/miso-lims/blob/78d3c908bdf72379583faedc8d74dc1bd62ab962/miso-web/src/main/webapp/scripts/hot.js#L341): Add the "let", "const" or "var" keyword to this declaration of "prop" to make it explicit. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2703)
1. ![BLOCKER][BLOCKER] [hot.js#L386](https://github.com/TGAC/miso-lims/blob/78d3c908bdf72379583faedc8d74dc1bd62ab962/miso-web/src/main/webapp/scripts/hot.js#L386): Add the "let", "const" or "var" keyword to this declaration of "needsRender" to make it explicit. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2703)
1. ![BLOCKER][BLOCKER] [hot.js#L1013](https://github.com/TGAC/miso-lims/blob/78d3c908bdf72379583faedc8d74dc1bd62ab962/miso-web/src/main/webapp/scripts/hot.js#L1013): Add the "let", "const" or "var" keyword to this declaration of "HotTarget" to make it explicit. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2703)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L549](https://github.com/TGAC/miso-lims/blob/78d3c908bdf72379583faedc8d74dc1bd62ab962/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L549): Define a constant instead of duplicating this literal "title" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L576](https://github.com/TGAC/miso-lims/blob/78d3c908bdf72379583faedc8d74dc1bd62ab962/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L576): Define a constant instead of duplicating this literal "library" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L610](https://github.com/TGAC/miso-lims/blob/78d3c908bdf72379583faedc8d74dc1bd62ab962/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L610): Define a constant instead of duplicating this literal "Libraries" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L818](https://github.com/TGAC/miso-lims/blob/78d3c908bdf72379583faedc8d74dc1bd62ab962/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L818): Define a constant instead of duplicating this literal "Dilutions" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L833](https://github.com/TGAC/miso-lims/blob/78d3c908bdf72379583faedc8d74dc1bd62ab962/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L833): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L856](https://github.com/TGAC/miso-lims/blob/78d3c908bdf72379583faedc8d74dc1bd62ab962/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L856): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L918](https://github.com/TGAC/miso-lims/blob/78d3c908bdf72379583faedc8d74dc1bd62ab962/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L918): Define a constant instead of duplicating this literal "_POOL" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'
[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**